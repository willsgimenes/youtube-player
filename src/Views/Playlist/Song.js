import React from 'react'
import Song from '../../components/Song'

import { bindActionCreators } from 'redux'
import { connect } from 'react-redux'
import { changeSong } from '../../actions'

function mapStateToProps ({ player: { song } }) {
  return {
    song
  }
}

function mapDispatchToProps (dispatch) {
  return bindActionCreators({
    changeSong
  }, dispatch)
}

class SongComponent extends React.Component {
  constructor (props) {
    super(props)
    this.LoadMusic = this.LoadMusic.bind(this)
    this.getSongData = this.getSongData.bind(this)
  }

  LoadMusic (song) {
    const { changeSong } = this.props
    changeSong(song.id)
  };

  getSongData (item, index) {
    const { song } = this.props
    let songItem

    Object.keys(song).forEach(function (key) {
      if (song[key].id === item) songItem = song[key]
    })

    const {
      id,
      title,
      song: {
        id: songId,
        type,
        duration,
        prettyDuration,
        author,
        url,
        cleanTitle
      }
    } = songItem

    return {
      id,
      title,
      songId,
      type,
      duration,
      prettyDuration,
      author,
      url,
      cleanTitle,
      item,
      index
    }
  }

  render () {
    const {
      item,
      index,
      song
    } = this.props

    const actions = {
      handleClick: this.LoadMusic
    }

    return song ? <Song {...this.getSongData(item, index)} {...actions} /> : ''
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(SongComponent)
