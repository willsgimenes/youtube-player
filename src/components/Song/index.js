import React from 'react'
import truncateString from '../../Helpers/Truncate'
import PropTypes from 'prop-types'

import styled from 'styled-components'

const Wrapper = styled.li`
  background: ${props => props.active ? '#eaeaea' : '#fff'}
  margin: 5px 0 5px 10px;
  clear: both;
  list-style: none;
  cursor: pointer;
  display: flex;
  height: 79px;
`
const Thumbnail = styled.div`
  width: 40%;
  float: left;
  background: #ccc;
  position: relative;
  overflow: hidden;
`

const Img = styled.img`
  width: 100%;
  z-index: 1;
`

const Duration = styled.div`
  background: rgba(0, 0, 0, .7);
  color: #fff;
  bottom: 3px;
  right: 3px;
  position: absolute;
  z-index: 2;
  font-size: .9em;
  padding: 3px 4px;
  border-radius: 4px;
  -webkit-font-smoothing: antialiased;
`

const SongInfo = styled.div`
  padding: 0 5px;
`

const SongTitle = styled.div`
  font-weight: bold;
  width: 100%;
  position: relative;
  margin-bottom: 10px;
`

const SongAuthor = styled.div`
  margin-bottom: 10px;
`

const SongText = styled.div`
  font-size: 1.1em;
  margin: 0;
  width: 165px;
  line-height: 1.4em;

`

class Song extends React.Component {
  render () {
    const {
      item,
      index,
      id,
      title,
      songId,
      prettyDuration,
      author,
      handleClick
    } = this.props

    return (
      <Wrapper
        active={index === item}
        key={id}
        onClick={() => handleClick(this.props)}
        data-cy={`song-${id}`}
        data-active={index === item}
      >
        <Thumbnail>
          <Img src={`https://img.youtube.com/vi/${songId}/default.jpg`} alt='' />
          <Duration>
            {prettyDuration}
          </Duration>
        </Thumbnail>

        <SongInfo>
          <SongTitle>
            <SongText>{truncateString(title, 40)}</SongText>
          </SongTitle>
          <SongAuthor>
            {author}
          </SongAuthor>
        </SongInfo>
      </Wrapper>
    )
  }
}

Song.propTypes = {
  handleClick: PropTypes.func.isRequired
}

export default Song
