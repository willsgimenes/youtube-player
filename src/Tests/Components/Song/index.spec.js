/* global describe */
/* global beforeEach */
/* global it */
/* global expect */
/* global jest */

import React from 'react'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Song from '../../../components/Song'

import toJson from 'enzyme-to-json'
import 'jest-styled-components'

Enzyme.configure({ adapter: new Adapter() })

describe('<Song />', () => {
  let
    wrapper,
    initialState

  beforeEach(() => {
    initialState = {
      item: 1,
      index: 1,
      id: 1,
      title: 1,
      songId: 1,
      prettyDuration: 1,
      author: 1,
      handleClick: jest.fn()
    }

    wrapper = mount(<Song {...initialState} />)
  })

  it('should render properly', () => {
    expect(wrapper.find('li')).toHaveLength(1)
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should have song--active class', () => {
    initialState = {
      item: 1,
      index: 1,
      id: 1,
      title: 1,
      songId: 1,
      prettyDuration: 1,
      author: 1,
      handleClick: jest.fn()
    }

    wrapper = shallow(<Song {...initialState} />)

    expect(wrapper).toHaveStyleRule('background', '#eaeaea')
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should test handleClick function', () => {
    initialState = {
      item: 1,
      index: { id: 1 },
      id: 1,
      title: 1,
      songId: 1,
      prettyDuration: 1,
      author: 1,
      handleClick: jest.fn()
    }

    wrapper = shallow(<Song {...initialState} />)

    const instance = wrapper.instance()
    const spy = jest.spyOn(instance.props, 'handleClick')
    instance.forceUpdate()
    wrapper.dive().find('li').simulate('click')
    expect(spy).toHaveBeenCalled()
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})
