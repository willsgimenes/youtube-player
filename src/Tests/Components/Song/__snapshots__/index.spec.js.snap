// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Song /> should have song--active class 1`] = `
<styled.li
  active={true}
  data-active={true}
  data-cy="song-1"
  key="1"
  onClick={[Function]}
>
  <styled.div>
    <styled.img
      alt=""
      src="https://img.youtube.com/vi/1/default.jpg"
    />
    <styled.div>
      1
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div>
        1
      </styled.div>
    </styled.div>
    <styled.div>
      1
    </styled.div>
  </styled.div>
</styled.li>
`;

exports[`<Song /> should render properly 1`] = `
.c0 {
  background: #eaeaea;
  margin: 5px 0 5px 10px;
  clear: both;
  list-style: none;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 79px;
}

.c1 {
  width: 40%;
  float: left;
  background: #ccc;
  position: relative;
  overflow: hidden;
}

.c2 {
  width: 100%;
  z-index: 1;
}

.c3 {
  background: rgba(0,0,0,.7);
  color: #fff;
  bottom: 3px;
  right: 3px;
  position: absolute;
  z-index: 2;
  font-size: .9em;
  padding: 3px 4px;
  border-radius: 4px;
  -webkit-font-smoothing: antialiased;
}

.c4 {
  padding: 0 5px;
}

.c5 {
  font-weight: bold;
  width: 100%;
  position: relative;
  margin-bottom: 10px;
}

.c7 {
  margin-bottom: 10px;
}

.c6 {
  font-size: 1.1em;
  margin: 0;
  width: 165px;
  line-height: 1.4em;
}

<Song
  author={1}
  handleClick={[Function]}
  id={1}
  index={1}
  item={1}
  prettyDuration={1}
  songId={1}
  title={1}
>
  <styled.li
    active={true}
    data-active={true}
    data-cy="song-1"
    key="1"
    onClick={[Function]}
  >
    <li
      className="c0"
      data-active={true}
      data-cy="song-1"
      onClick={[Function]}
    >
      <styled.div>
        <div
          className="c1"
        >
          <styled.img
            alt=""
            src="https://img.youtube.com/vi/1/default.jpg"
          >
            <img
              alt=""
              className="c2"
              src="https://img.youtube.com/vi/1/default.jpg"
            />
          </styled.img>
          <styled.div>
            <div
              className="c3"
            >
              1
            </div>
          </styled.div>
        </div>
      </styled.div>
      <styled.div>
        <div
          className="c4"
        >
          <styled.div>
            <div
              className="c5"
            >
              <styled.div>
                <div
                  className="c6"
                >
                  1
                </div>
              </styled.div>
            </div>
          </styled.div>
          <styled.div>
            <div
              className="c7"
            >
              1
            </div>
          </styled.div>
        </div>
      </styled.div>
    </li>
  </styled.li>
</Song>
`;

exports[`<Song /> should test handleClick function 1`] = `
<styled.li
  active={false}
  data-active={false}
  data-cy="song-1"
  key="1"
  onClick={[Function]}
>
  <styled.div>
    <styled.img
      alt=""
      src="https://img.youtube.com/vi/1/default.jpg"
    />
    <styled.div>
      1
    </styled.div>
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.div>
        1
      </styled.div>
    </styled.div>
    <styled.div>
      1
    </styled.div>
  </styled.div>
</styled.li>
`;
